### ISSUE TITLE
support for non-seekable input

### ISSUE THREAD

**Issue by umlaeute** [libsndfile/libsndfile#280]
(this was originally reported as https://bugs.debian.org/#761307)

Support for non-seekable inputs is questionable:

~~~sh
$ head -c 10000 /dev/zero | oggenc -r - -o- | sndfile-convert -pcm16 /dev/stdin test.wav
Not able to open input file /dev/stdin.
Internal error : SF_INFO struct incomplete.

$ cat somefile.ogg | sndfile-convert -pcm16 /dev/stdin test.wav
Not able to open input file /dev/stdin.
Supported file format but file is malformed.
~~~

strace shows (in both cases) that it tries to seek although it
(correctly in this case) determined that pipes are not seekable.

That's for Ogg/Vorbis, but grepping the code shows, for various
formats, a number of calls of `psf_fseek` which are not from the
respective seek methods, so would be called for non-seekable files.

I see no easy fix except buffering the input internally. So perhaps
it's best to declare that the library simply doesn't work with
non-seekable inputs and remove the code that pretends it does,
shifting the burden of buffering non-seekable inputs to the user of
the library (which I will now do in my own code via
`sf_open_virtual`).

**Comment by tgarc** [comment]
I also ran into this issue when using the virtual_io functions. libsndfile seems to assume that virtual io files are always seekable.

**Comment by erikd** [comment]
This is not so much a libsndfile limitation as a limitation of the file formats that libsndfile is trying to read/write. For instance to write a well formed WAVE file *requires* that during `sf_close` that a value (that depends on the amount of audio data written) is written to the start of the file. That requires a seek. 

The only way that could possibly be avoided is if the libsndfile API allowed the caller to specify the number of audio frames that were going to be written up-front during `sf_open`, but supporting this change would require a huge amount of work and is unlikely to happen.

**Comment by tgarc** [comment]
True. What about in read only situations? I only care about reading a non-seekable file object, are there similar seek requirements for that case?

**Comment by erikd** [comment]
@tgarc Specific examples (either using the API or programs like `sndfile-info` and `sndfile-convert`) would help. Looking back at the original issue that was raised, I am inclined to agree that:
```
head -c 10000 /dev/zero | oggenc -r - -o- | sndfile-convert -pcm16 - test.wav
```
should work even if it doesn't right now.

Unfortunately, this cannot be "fixed" globally. It needs fixing on a format-by-format basis.
